### PR TITLE
Create JSONFieldWrapper to support JSONField shipped by Django and jsonfield

### PR DESCRIPTION
**Pull request by khchine5**

### CODE REVIEW

File: shop/models/fields.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from django.db import connection
+
+POSTGRES_FALG = False
```
Comment (rfleschenberg) [inline shop/models/fields.py]: What is the meaning of `FALG` here?

**Comment by rfleschenberg** [comment]
Hi @khchine5. Thanks for the PR!

How will this affect shops that are already deployed? Will users have to take any special steps to update? If yes, we should document these steps.

File: shop/models/fields.py (hunk 1 above)
Comment (khchine5) [inline shop/models/fields.py]: Ah , I am sorry.It is a misspelled variable.
Fixed.

File: shop/models/fields.py [hunk 2]
```diff
@@ -12,6 +12,8 @@
     import psycopg2
 
     version = re.search('([0-9.]+)', psycopg2.__version__ or "0").group(0)
+    # To be able to use the Django version of JSONField, it requires to have PostgreSQL ≥ 9.4 , otherwise some issues
+    #  could be faced.
     if POSTGRES_FLAG and (LooseVersion(version) >= LooseVersion('2.5.4')):
```
Comment (rfleschenberg) [inline shop/models/fields.py]: This looks like it checks the psycopg version, not the version of Postgres itself. Or are these two somehow related?

**Comment by rfleschenberg** [comment]
The related issue is #398.

File: shop/models/fields.py (hunk 2 above)
Comment (khchine5) [inline shop/models/fields.py]: Yes. Actually the new jsonfield shipped with django required postgres >=
9.4 and psycopy2 >= 2.5.4 .since I haven't found a way to check
postgresql's version ,I plan to add it in the documentation.
https://docs.djangoproject.com/en/1.10/ref/contrib/postgres/fields/#jsonfield

**Comment by jrief** [comment]
Hi @khchine5 
Thanks for you pull request, I just looked at it.
Unfortunately I can not merge it, because of some issues.
I'll add comments to the corresponding code parts.

File: shop/models/fields.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from distutils.version import LooseVersion
+import re
+import subprocess
+from django.db import connection
+
+POSTGRES_FLAG = False
```
Comment (jrief) [inline shop/models/fields.py]: I would not use a flag here, evaluating for the vendor string is enough

File: shop/models/fields.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+if str(connection.vendor) == 'postgresql':
+    POSTGRES_FLAG = True
+try:
+    if POSTGRES_FLAG:
+        import psycopg2
+
+        psycopg2_version = re.search('([0-9.]+)', psycopg2.__version__ or "0").group(0)
```
Comment (jrief) [inline shop/models/fields.py]:
```
psycopg2.__version__.split('.')
```
is a better way to check for versions

File: shop/models/fields.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        try:
+            process = subprocess.Popen(
```
Comment (jrief) [inline shop/models/fields.py]: Never ever run shell command from within Django. This is very unportable and dangerous.
A better way to achieve this is to use raw SQL:
```
with connection.cursor() as cursor:
    cursor.execute("SELECT version()")
    row = cursor.fetchone()
```
``row`` then contains the version you can further process.

File: shop/models/fields.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        else:
+            raise ImportError
+    else:
+        raise ImportError
+
+except ImportError:
+    from jsonfield.fields import JSONField
+
+
+class JSONFieldWrapper(JSONField):
```
Comment (jrief) [inline shop/models/fields.py]: Since we want to wrap that field, I would overwrite its classname, ie.
```
class JSONField(JSONField):
    ...
```

File: shop/models/fields.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    def __init__(self, *args, **kwargs):
+        kwargs.update({'default': {}})
```
Comment (jrief) [inline shop/models/fields.py]: since you set the default value, you must also override the method ``deconstruct ``, otherwise we get problems when loading fixtures. This is explained here: https://docs.djangoproject.com/en/1.10/howto/custom-model-fields/#field-deconstruction

**Comment by khchine5** [comment]
Hi @jrief .
Thanks for all your comments.

File: shop/models/fields.py (hunk 4 above)
Comment (khchine5) [inline shop/models/fields.py]: Yes.
Using 'get_tuple_version()' from shop.apps could be a good option.

File: shop/models/fields.py (hunk 3 above)
Comment (khchine5) [inline shop/models/fields.py]: Yes. I thought I may need it later but I was wrong.

File: shop/models/fields.py (hunk 5 above)
Comment (khchine5) [inline shop/models/fields.py]: Yes.

**Comment by jrief** [comment]
Thanks for this pull request.
I merged it into my own develop branch. It should become part of the official branch as soon as I merge them.

@rfleschenberg could you please check in your Postgres environment.

**Comment by rfleschenberg** [comment]
The PR still has an issue:

```
(acceed-shop) rene@rex ~/projekte/acceed2 (master) $ ./manage.py migrate
Traceback (most recent call last):
  File "./manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/home/rene/.virtualenvs/acceed-shop/lib/python3.5/site-packages/django/core/management/__init__.py", line 353, in execute_from_command_line
    utility.execute()
  File "/home/rene/.virtualenvs/acceed-shop/lib/python3.5/site-packages/django/core/management/__init__.py", line 327, in execute
    django.setup()
  File "/home/rene/.virtualenvs/acceed-shop/lib/python3.5/site-packages/django/__init__.py", line 18, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/home/rene/.virtualenvs/acceed-shop/lib/python3.5/site-packages/django/apps/registry.py", line 108, in populate
    app_config.import_models(all_models)
  File "/home/rene/.virtualenvs/acceed-shop/lib/python3.5/site-packages/django/apps/config.py", line 202, in import_models
    self.models_module = import_module(models_module_name)
  File "/home/rene/.virtualenvs/acceed-shop/lib/python3.5/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 986, in _gcd_import
  File "<frozen importlib._bootstrap>", line 969, in _find_and_load
  File "<frozen importlib._bootstrap>", line 958, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 673, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 665, in exec_module
  File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
  File "/home/rene/projekte/acceed2/acceed/models.py", line 15, in <module>
    from shop.models.defaults.address import BillingAddress, ShippingAddress  # noqa
  File "/home/rene/.virtualenvs/acceed-shop/lib/python3.5/site-packages/shop/models/__init__.py", line 2, in <module>
    from .notification import Notification, NotificationAttachment
  File "/home/rene/.virtualenvs/acceed-shop/lib/python3.5/site-packages/shop/models/notification.py", line 18, in <module>
    from .customer import CustomerModel
  File "/home/rene/.virtualenvs/acceed-shop/lib/python3.5/site-packages/shop/models/customer.py", line 18, in <module>
    from shop.models.fields import JSONField
  File "/home/rene/.virtualenvs/acceed-shop/lib/python3.5/site-packages/shop/models/fields.py", line 20, in <module>
    if (psycopg2_version[0]) >= (2, 5, 4) and (postgres_version >= (9, 4)):
TypeError: unorderable types: int() >= tuple()
(acceed-shop) rene@rex ~/projekte/acceed2 (master) $ 
```

**Comment by rfleschenberg** [comment]
Looks like the above problem is limited to Python 3. @khchine5, can you test your PR changes on Python 3 and make them compatible?

**Comment by khchine5** [comment]
Hi all,
Sorry , I was busy last week.
Thanks @jrief .
Okay @rfleschenberg . It seems that you have already fixed the above problem. But I got an other problem with Python3 which it is not related to this merge but with PolymorphicManager. I will work in the bonus points ;)

**Comment by rfleschenberg** [comment]
Thanks, that would be great!

**Comment by khchine5** [comment]
@rfleschenberg , What you meant by making testcases which fails with Python3 . I don't see the point.

**Comment by rfleschenberg** [comment]
It would be good to have test coverage for this code, so we get alerted if something breaks on Python 3 before a user hits it. But I will do this as part of the Postgres tests -- never mind. Thanks!
